lib/xnvme_be_macos_driverkit_mem.cpp

```diff
@@ -0,0 +1,87 @@
+// SPDX-FileCopyrightText: Samsung Electronics Co., Ltd
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifdef XNVME_BE_MACOS_ENABLED
+#include <xnvme_be_macos_driverkit.h>
+#endif
+
+extern "C" {
+
+#include <libxnvme.h>
+#include <xnvme_be.h>
+#include <xnvme_be_nosys.h>
+#ifdef XNVME_BE_MACOS_ENABLED
+
+#include <mach/mach_error.h>
+
+#include <xnvme_dev.h>
+#include <errno.h>
+#include <sys/mman.h>
+
+void *
+xnvme_be_macos_driverkit_buf_alloc(const struct xnvme_dev *dev, size_t nbytes,
+				   uint64_t *XNVME_UNUSED(phys))
+{
+	uint64_t token;
+	struct xnvme_be_macos_driverkit_state *state = (struct xnvme_be_macos_driverkit_state *)dev->be.state;
+	long sz = sysconf(_SC_PAGESIZE);
+
+	if (sz == -1) {
+		XNVME_DEBUG("FAILED: sysconf(), errno: %d", errno);
+		return NULL;
+	}
+	XNVME_DEBUG("Page size: %lx", sz);
+
+	// nbytes has to be a multiple of alignment. Therefore, we round up to the nearest
+	// multiple.
+	XNVME_DEBUG("Unaligned size: %lx", nbytes);
+	nbytes = (1 + ((nbytes - 1) / sz)) * (sz);
+	XNVME_DEBUG("Aligned size: %lx", nbytes);
+
+	void* ptr = xnvme_buf_virt_alloc(sz, nbytes);
+	size_t output_cnt = sizeof(token);
+	*(uint64_t*) ptr = nbytes; // Put actual size inside allocated memory. Driver on the other end doesn't correctly get this.
```

Comment:
Can't this be part of a struct used with `IOConnectCallStructMethod()` instead of passing the pointer directly?

I.e.,

```c
struct alloc_buffer_request {
  void *ptr;
  size_t len;
};

IOConnectCallStructMethod(..., NVME_ALLOC_BUFFER, req, 1, &token, &output_cnt);
```

As I understand it, the remote side (MacVFN) knows what structs to expect, so it can unpack it nicely. Or have I misunderstood the protocol?

Reply:
On the MacVFN side, I see that you are retrieving the argument directly as an `IOMemoryDescriptor`. What governs when what is passed is just an `OSData` and when it is an `IOMemoryDescriptor`? The documentation is a little vague on that point... ;)

Reply:
Mads explained offline. No, it is not trivial to just pass a nice "parameter structure" due to the way IOKit maps the arguments. Bottomline is that you just have to suck it up and pack it like above.